File: CODE_SRC/src2/module/stu.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "../include/config.h"
#include "../include/stu.h"
#include "../include/tools.h"
#include "../include/vars.h"

//初始化
void init(){
    FILE *fp;
    //打开文件
    printf("Now opening the file.....%s\n",FILENAME);
    if( (fp=fopen(FILENAME, "rb+")) == NULL && (fp=fopen(FILENAME, "wb+")) == NULL ){
        // pause("Error on open %s file!", FILENAME);
        printf("Error on open %s file!\n", FILENAME);
        exit(EXIT_FAILURE);
    }

    //获取文件长度
    long fileSize = getFileSize(fp);
    printf("The size of file:%ld\n", fileSize);
}


// 提示用户按键继续的函数
void pause(const char *message) {
    printf("%s\n", message);
    getchar();  // 等待用户按任意键继续
}

// 读取学生信息并返回成功读取的学生数量
int readStudentData(STU students[], int max_students) {
    FILE *fp = fopen(FILENAME, "r");
    if (fp == NULL) {
        perror("Failed to open file");
        return 0;
    }

    char line[1024];
    int count = 0;

    // 跳过第一行表头
    if (fgets(line, sizeof(line), fp) == NULL) {
        printf("Failed to read the header.\n");
        fclose(fp);
        return 0;
    }

    // 读取每一行学生数据
    while (fgets(line, sizeof(line), fp) && count < max_students) {
        STU stu;

        // 使用 sscanf 解析每一行
        if (sscanf(line, "%d,%63[^,],%9[^,],%d,%f,%f,%f", 
            &stu.id, stu.name, stu.sex, &stu.age, &stu.math, &stu.cn, &stu.en) == 7) {
            students[count++] = stu;
        } else {
            // 如果某一行格式不正确，打印错误
            printf("Error reading line: %s\n", line);
        }
    }

    fclose(fp);
    return count;
}

// 显示所有学生信息
void showAllStu() {
    STU students[100];  // 假设最多有100个学生
    int student_count = readStudentData(students, 100);

    if (student_count == 0) {
        pause("没有学生信息可显示！按任意键返回...");
        return;
    }

    // 打印表头
    printf("ID\tName\tSex\tAge\tMath\tChinese\tEnglish\n");
    printf("-----------------------------------------------------\n");

    // 打印所有学生信息
    for (int i = 0; i < student_count; i++) {
        printf("%d\t%s\t%s\t%d\t%.1f\t%.1f\t%.1f\n", 
            students[i].id, students[i].name, students[i].sex, students[i].age, 
            students[i].math, students[i].cn, students[i].en);
    }
}

// 根据成绩查询学生信息
void findStuByScores() {
    // 设定成绩查询的范围
    float min_score, max_score;

    // 提示用户输入成绩范围
    printf("Enter the minimum score:");
    scanf("%f", &min_score);
    printf("Enter the maximum score:");
    scanf("%f", &max_score);

    STU students[100];  // 假设最多有100个学生
    int student_count = readStudentData(students, 100);

    if (student_count == 0) {
        pause("No student information to query! Press any key to return..");
        return;
    }

    int found = 0;  // 用于标记是否找到符合条件的学生

    // 打印表头
    printf("ID\tName\tSex\tAge\tMath\tChinese\tEnglish\n");
    printf("-----------------------------------------------------\n");

    // 查找符合条件的学生
    for (int i = 0; i < student_count; i++) {
        if (students[i].math >= min_score && students[i].math <= max_score) {
            printf("%d\t%s\t%s\t%d\t%.1f\t%.1f\t%.1f\n", 
                students[i].id, students[i].name, students[i].sex, students[i].age, 
                students[i].math, students[i].cn, students[i].en);
            found = 1;
        }
    }

    // 如果没有找到符合条件的学生，提示错误信息
    if (!found) {
        pause("No student information to query! Press any key to return...");
    }
}

File: src/module/stu.rs
use std::io::{self, BufRead, BufReader, Write};
use std::fs::File;

// 在文件顶部定义一个常量
const FILE_PATH: &str = "./data/stu.csv";

pub struct Stu {
    pub id: i32,
    pub name: String,
    pub sex: String,
    pub age: i32,
    pub math: f32,
    pub cn: f32,
    pub en: f32,
}

pub fn init() {
    // 确保目录存在
    std::fs::create_dir_all("./data").expect("无法创建目录");
    
    if !std::path::Path::new(FILE_PATH).exists() {
        let mut file = std::fs::File::create(FILE_PATH).expect("无法创建文件");
        // 写入表头
        writeln!(file, "ID,Name,Sex,Age,Math,Chinese,English").expect("无法写入表头");
        println!("创建了新的学生数据文件");
    }
    let metadata = std::fs::metadata(FILE_PATH).expect("无法获取文件元数据");
    println!("文件大小：{} 字节", metadata.len());
}

pub fn pause(message: &str) {
    println!("{}", message);
    let _ = std::io::stdin().read_line(&mut String::new());
}

pub fn read_student_data() -> Result<Vec<Stu>, Box<dyn std::error::Error>> {
    // 使用相同的常量
    let file = std::fs::File::open(FILE_PATH)?;
    let reader = std::io::BufReader::new(file);

    let mut students = Vec::new();

    let mut lines = reader.lines();
    if let Some(Ok(_)) = lines.next() {}

    for line_result in lines {
        let line = line_result?;
        let parts: Vec<&str> = line.split(',').collect();
        if parts.len() != 7 {
            eprintln!("无效的行格式：{}", line);
            continue;
        }

        let id = parts[0].parse::<i32>()?;
        let name = parts[1].trim().to_string();
        if name.len() > 19 {
            eprintln!("姓名超过长度限制：{}", name);
            continue;
        }
        let sex = parts[2].trim().to_string();
        if sex.len() > 9 {
            eprintln!("性别字段过长：{}", sex);
            continue;
        }
        let age = parts[3].parse::<i32>()?;
        let math = parts[4].parse::<f32>()?;
        let cn = parts[5].parse::<f32>()?;
        let en = parts[6].parse::<f32>()?;

        students.push(Stu {
            id,
            name,
            sex,
            age,
            math,
            cn,
            en,
        });
    }

    Ok(students)
}

pub fn show_all_stu() {
    match read_student_data() {
        Ok(students) => {
            if students.is_empty() {
                pause("没有学生信息可显示。按任意键继续...");
                return;
            }
            println!(
                "ID\tName\tSex\tAge\tMath\tChinese\tEnglish"
            );
            println!("{}", "-----------------------------------------------------");
            for stu in students.iter() {
                println!(
                    "{}\t{}\t{}\t{}\t{:.1}\t{:.1}\t{:.1}",
                    stu.id, stu.name, stu.sex, stu.age, stu.math, stu.cn, stu.en
                );
            }
            pause("按任意键继续...");
        }
        Err(e) => {
            eprintln!("读取学生数据时出错: {}", e);
            pause("错误发生。按任意键继续...");
        }
    }
}

pub fn find_stu_by_scores() {
    println!("输入最低数学成绩：");
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    let min_score: f32 = input.trim().parse().expect("请输入有效数字");

    println!("输入最高数学成绩：");
    input.clear();
    std::io::stdin().read_line(&mut input).unwrap();
    let max_score: f32 = input.trim().parse().expect("请输入有效数字");

    match read_student_data() {
        Ok(students) => {
            let mut found = false;
            println!(
                "ID\tName\tSex\tAge\tMath\tChinese\tEnglish"
            );
            println!("{}", "-----------------------------------------------------");
            for stu in students.iter() {
                if (stu.math >= min_score) && (stu.math <= max_score) {
                    println!(
                        "{}\t{}\t{}\t{}\t{:.1}\t{:.1}\t{:.1}",
                        stu.id, stu.name, stu.sex, stu.age, stu.math, stu.cn, stu.en
                    );
                    found = true;
                }
            }
            if !found {
                pause("没有找到符合条件的学生。按任意键继续...");
            } else {
                pause("按任意键继续...");
            }
        }
        Err(e) => {
            eprintln!("读取学生数据时出错: {}", e);
            pause("错误发生。按任意键继续...");
        }
    }
}